src/ColinHDev/CPlot/provider/DataProvider.php

```diff
@@ -1299,4 +1313,213 @@ public function savePlotRate(Plot $plot, PlotRate $plotRate) : Generator {
+                            null,
+                            $XUID,
+                            $playerName
+                        );
+                    }
+
+                    $senderData = new PlotPlayer($playerData, PlotPlayer::STATE_DENIED);
+                    $plot->addPlotPlayer($senderData);
+                    yield from DataProvider::getInstance()->savePlotPlayer($plot, $senderData);
+                }
+
+                //load common flags
+                $flag = Flags::PVP()->createInstance($record["pvp"]);
+                $flag = $plot->getLocalFlagByID($flag->getID())?->merge($flag->getValue()) ?? $flag;
+                $plot->addFlag($flag);
+                $this->savePlotFlag($plot, $flag);
+            }
+            foreach($mergeRecords as $mergeRecord) {
+                // load world
+                /** @var WorldSettings|false $world */
+                $world = yield $this->awaitWorld($mergeRecord["level"]);
+                if($world === false)
+                    continue;
+
+                // load merge plot 1
+                /** @var Plot|null $plot */
+                $plot = yield $this->awaitPlot($mergeRecord["level"], (int)$mergeRecord["originX"], (int)$mergeRecord["originZ"]);
+                if($plot === null)
+                    continue;
+
+                // load merge plot 2
+                /** @var Plot|null $plotToMerge */
+                $plotToMerge = yield $this->awaitPlot($mergeRecord["level"], (int)$mergeRecord["mergedX"], (int)$mergeRecord["mergedZ"]);
+                if($plotToMerge === null)
+                    continue;
+
+                // complete merge logic
+                yield from DataProvider::getInstance()->awaitPlotDeletion($plotToMerge);
+                foreach($plotToMerge->getMergePlots() as $mergePlot){
+                    $plot->addMergePlot($mergePlot);
+                    yield from $this->addMergePlot($plot, $mergePlot);
+                }
+                foreach($plotToMerge->getPlotPlayers() as $mergePlotPlayer) {
+                    $plot->addPlotPlayer($mergePlotPlayer);
+                    yield from $this->savePlotPlayer($plot, $mergePlotPlayer);
+                }
+                foreach ($plotToMerge->getFlags() as $mergeFlag) {
+                    $flag = $plot->getFlagByID($mergeFlag->getID());
+                    if ($flag === null) {
+                        $flag = $mergeFlag;
+                    } else {
+                        $flag = $flag->merge($mergeFlag->getValue());
+                    }
+                    $plot->addFlag($flag);
+                    yield from DataProvider::getInstance()->savePlotFlag($plot, $flag);
+                }
+                foreach ($plotToMerge->getPlotRates() as $mergePlotRate) {
+                    $plot->addPlotRate($mergePlotRate);
+                    yield from DataProvider::getInstance()->savePlotRate($plot, $mergePlotRate);
+                }
```

Comment:
So, I looked into how MyPlot stores data of merged plots. Please correct me if I'm wrong, you know your plugin better than I do:
Contrary to CPlot, MyPlot not only stores the merge information in the `mergedPlotsV2` table but also the general plot information in the `plotsV2` table.
Also, as in CPlot, if two or more plots are merged, there is only one origin plot, which is 

If that is the case, couldn't we first loop over the merge records and organize them, so when we loop over the plot records, we are able to filter those out who are merged anyway?
I don't really like that we add plots to the database just to immediately delete them.